Handle peer event async in metrics

cammellos opened the pull request:
There might be an issue on how we handle metrics, which causes the p2p
server to hang.

updateNodeMetrics calls  a method on the p2p server, which
blocks until the server is available:

https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L301
https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L746
If there's back-pressure on the peer event feed
https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L783

The event channel above might become while updateNodeMetrics
is called, which means is never consumed, the server blocks on publishing on
it, and the two will deadlock (server waits for the channel above to be consumed,
this code waits for the server to respond to peerCount, which is in the same
event loop).

Calling it in a different go-routine will allow this code to keep
processing peer added events, therefore the server will not lock and keep processing requests.

cammellos committed:
Handle peer event async in metrics

There might be an issue on how we handle metrics, which causes the p2p
server to hang.

updateNodeMetrics calls  a method on the p2p server, which
blocks until the server is available:

https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L301
https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L746
If there's back-pressure on the peer event feed
https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L783

The event channel above might become while updateNodeMetrics
is called, which means is never consumed, the server blocks on publishing on
it, and the two will deadlock (server waits for the channel above to be consumed,
this code waits for the server to respond to peerCount, which is in the same
event loop).

Calling it in a different go-routine will allow this code to keep
processing peer added events, therefore the server will not lock and keep processing requests.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (3)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | ef06298f | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2139/1/) | 2021-02-01 09:50:32 | ~2 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210201-094814-ef0629.zip) |
| :heavy_check_mark: | ef06298f | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2139/1/) | 2021-02-01 09:52:43 | ~4 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210201-094814-ef0629.zip) |
| :heavy_check_mark: | ef06298f | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2139/1/) | 2021-02-01 09:56:51 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210201-094811-ef0629.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d9b13e68 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2139/2/) | 2021-02-01 13:45:59 | ~3 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210201-134258-d9b13e.zip) |
| :heavy_check_mark: | d9b13e68 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2139/2/) | 2021-02-01 13:47:00 | ~4 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210201-134256-d9b13e.zip) |
| :heavy_check_mark: | d9b13e68 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2139/2/) | 2021-02-01 13:49:00 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210201-134254-d9b13e.aar) |
| | | | | | | |
| :heavy_check_mark: | 45b64e6b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2139/3/) | 2021-02-01 13:47:53 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210201-134602-45b64e.zip) |
| :heavy_check_mark: | 45b64e6b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2139/3/) | 2021-02-01 13:49:20 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210201-134702-45b64e.zip) |
| :heavy_check_mark: | 45b64e6b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2139/3/) | 2021-02-01 13:53:45 | ~4 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210201-134902-45b64e.aar) |

jakubgs commented on metrics/node/subscribe.go:
```diff
@@ -31,9 +31,26 @@ func SubscribeServerEvents(ctx context.Context, node *node.Node) error {
+				// https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L746
+				// If there's back-pressure on the peer event feed
+				// https://github.com/status-im/status-go/blob/e60f425b45d00d3880b42fdd77b460ec465a9f55/vendor/github.com/ethereum/go-ethereum/p2p/server.go#L783
+				// The event channel above might become while updateNodeMetrics
```
"might become" what? Missing adjective.

jakubgs reviewed (approved):
Looks good to me, but do we really need all of that comment in there. I'd just do 1-3 sentences and put the full explanation in the commit. But to each their own.

cammellos committed:
Handle deadlock in peerpool

There was another deadlock in the peer pool.
Because we made the event handler asynchrnous, another deadlock popped
up, as the loop locks the global peerpool lock before processing events.
But the handlers also take the global look, effectively resulting in the
same situation we had before, i.e the loop is not running.

THE LOOP MUST BE RUNNING AT ALL TIMES OTHERWISE THE SERVER HANGS.

Samyoul commented on peers/peerpool.go:
```diff
@@ -103,6 +103,7 @@ type PeerPool struct {
 	cache  *Cache
 
 	mu                 sync.RWMutex
+	timeoutMu          sync.RWMutex
```
This is a good idea
